fhir/resources/core/fhirabstractmodel.py

```diff
@@ -47,66 +46,75 @@ def json_dumps(v, *, default, option=0, return_bytes=False):
 
 __author__ = "Md Nazrul Islam<email2nazrul@gmail.com>"
 
 logger = logging.getLogger(__name__)
 FHIR_COMMENTS_FIELD_NAME = "fhir_comments"
 
 
-class WrongResourceType(PydanticValueError):
+class WrongResourceType:
     code = "wrong.resource_type"
     msg_template = "Wrong ResourceType: {error}"
 
 
 class FHIRAbstractModel(BaseModel, abc.ABC):
     """Abstract base model class for all FHIR elements."""
 
+    json_loads = json_loads
+    json_dumps = json_dumps
     resource_type: str = ...  # type: ignore
+    model_config = ConfigDict(
+        extra="forbid",
+        populate_by_name=True,
+        validate_assignment=True,
+    )
 
     fhir_comments: typing.Union[str, typing.List[str]] = Field(
         None, alias="fhir_comments", element_property=False
     )
 
-    def __init__(__pydantic_self__, **data: typing.Any) -> None:
+    def __init__(self: "FHIRAbstractModel", **data: typing.Any) -> None:
         """ """
         resource_type = data.pop("resource_type", None)
-        errors = []
+        errors: typing.List[InitErrorDetails] = []
         if (
             "resourceType" in data
-            and "resourceType" not in __pydantic_self__.__fields__
+            and "resourceType" not in self.model_fields
         ):
             resource_type = data.pop("resourceType", None)
 
         if (
             resource_type is not None
-            and resource_type != __pydantic_self__.__fields__["resource_type"].default
+            and resource_type != self.model_fields["resource_type"].default
         ):
-            expected_resource_type = __pydantic_self__.__fields__[
+            expected_resource_type = self.model_fields[
                 "resource_type"
             ].default
-            error = (
-                f"``{__pydantic_self__.__class__.__module__}."
-                f"{__pydantic_self__.__class__.__name__}`` "
+            error_message = (
+                f"``{self.__class__.__module__}."
+                f"{self.__class__.__name__}`` "
                 f"expects resource type ``{expected_resource_type}``, "
                 f"but got ``{resource_type}``. "
                 "Make sure resource type name is correct and right "
                 "ModelClass has been chosen."
             )
-            errors.append(
-                ErrorWrapper(WrongResourceType(error=error), loc="resource_type")
+            init_error_details = InitErrorDetails(
+                type=WrongResourceType.code,
+                loc=("resource_type",),
+                ctx={"message": error_message},
+                input=resource_type
             )
+            errors.append(init_error_details)
         if errors:
-            raise ValidationError(errors, __pydantic_self__.__class__)
+            raise ValidationError(errors, self.__class__)
 
-        BaseModel.__init__(__pydantic_self__, **data)
+        BaseModel.__init__(self, **data)
 
     @classmethod
     def add_root_validator(
```

Comment:
In my opininon this is the most troublesome function. Its using the  ROOT_VALIDATOR_CONFIG_KEY which is deprecated in V2.  I've scoured the codebase and it is never used.

Frankly speaking I don't really understand the usecase that much. It looks like a custom way to add a root validator.

@nazrulworld I believe you coded it in as part of the initial dump.
Is this a feature you want to maintain or not?